### DISCUSSION TITLE
Rename "range" data types to "timestamps"

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/evm-monorepo#902] [Ideas]
Given we now have generic `createWithTimestamps` methods instead of the old variations (durations, deltas, ranges, milestones), shouldn't we rename the dedicated LockupLinear parameter for start / cliff / end from `range` to `timestamps` in the create method (and the `DataType` itself)?

CC: @sablier-labs/solidity

**Comment by PaulRBerg** [comment]
I was instinctually on board with this idea, but the following rationale made me change my mind:

- In the general context of Sablier Lockup, "range" is a different concept from "timestamps". It just so happens that in the case of LL, they are one and the same, but "timestamps" is a broader concept.
- In LD and LT
  - "timestamps" = start time + all segments/ tranches
  -  "range" = start time + end time
- Thus, changing `range` to `timestamps` in LL would mean that the `getRange` getter turns into `getTimestamps`
    - This would introduce inconsistency: LL would have `getTimestamps` and LD and LT `getRange`
    - And it would be a breaking change WRT V2.1, and we wish to minimize breaking changes (as per the rationale you shared elsewhere, can't remember the discussion)

Nonetheless, it's a valid point, and I suggest we sleep on it some more. We should also ask the auditors for feedback.

**Comment by smol-ninja** [comment]
I am in favour of this idea (I have also been thinking about this for a while and one more refactor I will start a separate discussion for). The `Range` struct contains information on timestamps. Even in the case of LD and LT, the range indicates unix timestamps for the segment/tranche start and end time. 

From [Cambridge dictionary](https://dictionary.cambridge.org/dictionary/english/range):

> Range is the amount, number, or type of something between an upper and a lower limit.

So in my opinion, it makes more sense to use range in the context of durations than the actual timestamps.

My comments/questions on @PaulRBerg's points:

> "timestamps" = start time + all segments/ tranches

Did you mean "timestamps" = start time + all segments/tranches timestamps? If so, that's more like a "current timestamp". I am not sure how it can be confused with the `Timestamp` struct.

> "range" = start time + end time

I didn't understand this as well. Can you please explain again? Did you mean range = start unix time + end unix time? If yes then how is it different from the above context?

> This would introduce inconsistency: LL would have getTimestamps and LD and LT getRange

We should also rename the range to timestamp in the context of Lockup Linear as well. In LL, the [range struct](https://github.com/sablier-labs/v2-core/blob/e080f20eafef0fc18049bcc77f1694db043860f1/src/types/DataTypes.sol#L260) contains three timestamps: start time, cliff time and end time. "Range" keyword can be used to determine stream range such as end time - start time, end time - cliff time and so on.

> And it would be a breaking change WRT V2.1, and we wish to minimize breaking changes

That is something I agree with as well.

**Comment by PaulRBerg** [comment]
Unfortunately, I did not understand most of your points, @smol-ninja. I think the crux of the misunderstanding is this:

- Range = start time unix and end time unix
- Timestamps = start time unix, all segment/ tranche timestamps, and end time unix

The code base considers these as two different concepts.

Thus, let us answer these questions in order to expedite the debate:

1. What structs do we want to modify? Let's be precise.
2. What variables do we want to modify? Again, let's be precise.
3. In LD and LT, do we want to merge the concepts of 'range' and 'timestamps' into a single concept ('timestamps')?
4. If yes, what should the `getTimestamps` getter return in LD and LT? Just the start time and the end time, or those two + all the timestamps in between?

---

Let me start:

1. All `Range` structs defined in `DataTypes` would become `Timestamp`.
2. All `range` variables used as function parameters in the create functions.
3. I'm undecided about this. I guess it would be easier to merge the concepts (noting that I changed my mind, when my previous view was that we should continue to keep them separated)
4. It should return just the start time and the end time so that we don't have to define another struct.

**Reply by smol-ninja** [reply]
Sorry, I initially misunderstood the term "range" as referring to the amount of time that passes (similar to "duration"). I thought they both refer to the same thing in English language yet we use them in different contexts. After having a conversation with ChatGPT, I now understand that "Range" specifically refers to timestamps rather than the length of time. So please disregard my previous comment.

Regarding the proposal to refactor "Range" to "Timestamp," I agree with the first two points you made.

3. Regarding the merge, how do you plan to combine them? Since LL would still require a different struct because of `cliff`. I suggest keeping them separate until we figure out a way standardize the `Range` (or new `Timestamp`) struct within the `Lockup` library 

5. Agreed.

**Reply by PaulRBerg** [reply]
Thank you for the follow-up, I understand you now.

3. There would be a `Timestamps` struct in each `Lockup*` model.

**Comment by PaulRBerg** [comment]
I suppose we don't have time to include this proposal in V2.2, do we?

**Reply by smol-ninja** [reply]
I think we can include this. Its just a refactor.

**Reply by PaulRBerg** [reply]
SGTM

**Reply by smol-ninja** [reply]
What should we call the variable?
1. `LockupDynamic.Timestamp memory timestamp`
2. `LockupDynamic.Timestamp memory timestampLD`

**Reply by PaulRBerg** [reply]
Simple `timestamp` should be fine

**Comment by smol-ninja** [comment]
Issue created:
- https://github.com/sablier-labs/v2-core/issues/911
